pandas/core/interval.py

```diff
@@ -0,0 +1,521 @@
+                                notnull, is_datetime_or_timedelta_dtype,
+                                is_integer_dtype, is_float_dtype)
+from pandas.core.index import (Index, _ensure_index, default_pprint,
+                               InvalidIndexError, MultiIndex)
+from pandas.lib import (Interval, IntervalMixin, IntervalTree,
+                        interval_bounds_to_intervals,
+                        intervals_to_interval_bounds)
+from pandas.util.decorators import cache_readonly
+import pandas.core.common as com
+
+
+_VALID_CLOSED = set(['left', 'right', 'both', 'neither'])
+
+
+def _get_next_label(label):
+    dtype = getattr(label, 'dtype', type(label))
+    if isinstance(label, (pd.Timestamp, pd.Timedelta)):
+        dtype = 'datetime64'
+    if is_datetime_or_timedelta_dtype(dtype):
+        return label + np.timedelta64(1, 'ns')
+    elif is_integer_dtype(dtype):
+        return label + 1
+    elif is_float_dtype(dtype):
+        return np.nextafter(label, np.infty)
+    else:
+        raise TypeError('cannot determine next label for type %r'
+                        % type(label))
+
+
+def _get_prev_label(label):
+    dtype = getattr(label, 'dtype', type(label))
+    if isinstance(label, (pd.Timestamp, pd.Timedelta)):
+        dtype = 'datetime64'
+    if is_datetime_or_timedelta_dtype(dtype):
+        return label - np.timedelta64(1, 'ns')
+    elif is_integer_dtype(dtype):
+        return label - 1
+    elif is_float_dtype(dtype):
+        return np.nextafter(label, -np.infty)
+    else:
+        raise TypeError('cannot determine next label for type %r'
+                        % type(label))
+
+
+def _get_interval_closed_bounds(interval):
+    """
+    Given an Interval or IntervalIndex, return the corresponding interval with
+    closed bounds.
+    """
+    left, right = interval.left, interval.right
+    if interval.open_left:
+        left = _get_next_label(left)
+    if interval.open_right:
+        right = _get_prev_label(right)
+    return left, right
+
+
+class IntervalIndex(IntervalMixin, Index):
+    """
+    Immutable Index implementing an ordered, sliceable set. IntervalIndex
+    represents an Index of intervals that are all closed on the same side.
+
+    .. versionadded:: 0.18
+
+    Properties
+    ----------
+    left, right : array-like (1-dimensional)
+        Left and right bounds for each interval.
+    closed : {'left', 'right', 'both', 'neither'}, optional
+        Whether the intervals are closed on the left-side, right-side, both or
+        neither. Defaults to 'right'.
+    name : object, optional
+        Name to be stored in the index.
+    """
+    _typ = 'intervalindex'
+    _comparables = ['name']
+    _attributes = ['name', 'closed']
+    _allow_index_ops = True
+    _engine = None # disable it
+
+    def __new__(cls, left, right, closed='right', name=None, fastpath=False):
+        # TODO: validation
+        result = IntervalMixin.__new__(cls)
+        result._left = _ensure_index(left)
```

Comment:
a lot of this should be in ``_simple_new`` (so you can reuse) (and then just call it from ``__new__``)